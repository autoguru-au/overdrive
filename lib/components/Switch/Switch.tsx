import cx from 'clsx';
import React, { FunctionComponent, memo, useCallback, useState } from 'react';
import styles from './style.scss';

export interface IProps {
	className?: string;
	disabled?: boolean;
	toggled?: boolean;

	onChange?(value: boolean): void;
}

const SwitchComponent: FunctionComponent<IProps> = ({
	className = '',
	disabled = false,
	toggled: incomingToggled = false,
	onChange,
}) => {
	const [toggled, setToggled] = useState<boolean>(incomingToggled);

	const [prevValue, setPrevValue] = useState<boolean>(incomingToggled);

	const onToggle = useCallback(() => {
		if (disabled) {
			return;
		}
		if (typeof onChange === 'function') {
			onChange(!toggled);
		}
		setToggled(!toggled);
	}, [incomingToggled, disabled, toggled]);

	if (prevValue !== incomingToggled) {
		setToggled(incomingToggled);
		setPrevValue(incomingToggled);
	}

	return (
		<div
			className={cx(
				styles.root,
				{
					[styles.toggled]: toggled,
				},
				className
			)}
			aria-disabled={disabled}
			aria-label="toggle value"
			onClick={onToggle}>
			<span className={styles.handle} />
		</div>
	);
};

export const Switch = memo(SwitchComponent);
