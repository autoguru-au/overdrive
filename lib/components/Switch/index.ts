export { Switch, IProps as IToggleButtonProps } from './Switch';
