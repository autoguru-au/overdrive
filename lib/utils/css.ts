const cssVarRegExp = /var\(([^)]+)\)/;

export const cssVarUnwrap = (value: string) => {
	const matches = cssVarRegExp.exec(value);

	return matches ? matches[1] : value;
};

export const getThemeTokenValue = (
	themeClass?: string | null,
	token?: string | null,
): string => {
	if (!themeClass || !token) return '';
	const cssVar = cssVarUnwrap(token);
	const themedElement = document.querySelector(`.${themeClass}`);
	if (!themedElement || !cssVar) return '';
	return (
		getComputedStyle(themedElement).getPropertyValue(cssVar)?.trim() || ''
	);
};
