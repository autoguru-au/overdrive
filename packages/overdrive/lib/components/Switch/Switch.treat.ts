import { style, styleMap } from 'treat';

const handleSize = '24px';
const handleOffset = '2px';
const borderSize = '1px';

export const root = style((theme) => ({
	fontSize: theme.typography.size['5'].fontSize,
	lineHeight: theme.typography.size['5'].fontSize,
	position: 'relative',
	width: `calc((2 * ${handleSize}) + ${handleOffset} - 2 * ${borderSize})`,
	height: `calc(${handleSize} + (${handleOffset} * 2))`,
	cursor: 'pointer',
	transition: 'background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
	border: `${borderSize} solid ${theme.colours.gamut.gray300}`,
	borderRadius: `calc(${handleSize} + ${handleOffset})`,
	backgroundColor: theme.colours.gamut.gray100,
}));

export const handleRoot = style((theme) => ({
	borderColor: theme.colours.gamut.gray200,
}));

export const handle = styleMap((theme) => ({
	default: {
		position: 'absolute',
		top: `calc(${handleOffset} - ${borderSize})`,
		left: `calc(${handleOffset} - ${borderSize})`,
		width: `${handleSize}`,
		height: `${handleSize}`,
		padding: 0,
		transition: 'transform 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
		color: 'white',
		border: `${borderSize} solid ${theme.colours.gamut.gray300}`,
		borderRadius: '50%',
		outline: 'none',
		backgroundColor: 'white',
		boxShadow: theme.elevation['2'],
		willChange: 'transform',
	},
	transition: {
		transform: `translateX(calc(${handleSize} - (2 * ${handleOffset})))`,
	},
}));

export const toggled = style((theme) => ({
	borderColor: theme.colours.gamut.green700,
	backgroundColor: theme.colours.gamut.green600,
}));

export const disabled = styleMap((theme) => ({
	default: {
		selectors: {
			'&[aria-disabled=true]': {
				cursor: 'not-allowed',
				borderColor: theme.colours.gamut.gray200,
				backgroundColor: theme.colours.gamut.gray100,
			},
		},
	},
	toggled: {
		selectors: {
			'&[aria-disabled=true]': {
				borderColor: theme.colours.gamut.green200,
				backgroundColor: theme.colours.gamut.green200,
			},
		},
	},
}));
