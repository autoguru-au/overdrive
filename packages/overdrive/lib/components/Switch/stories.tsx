import { action } from '@storybook/addon-actions';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import * as React from 'react';

import { Switch } from '.';

export default {
	title: 'Components/Switch',
	component: Switch,
} as ComponentMeta<typeof Switch>;

const Template: ComponentStory<typeof Switch> = ({
	...args
}) => (
	<Switch
		{...args}
	/>
);

const standardProps = {
	isDisabled: false,
	isSelected: false,
	onChange: action('onChange'),
	className: "toggleButton-class",
};

export const untoggled: ComponentStory<typeof Switch> =
	Template.bind(standardProps);
untoggled.args = standardProps;

const untoggledDisabledProps = {
	...standardProps,
	isDisabled: true,
};

export const untoggledDisabled: ComponentStory<typeof Switch> = Template.bind(
	untoggledDisabledProps,
);
untoggledDisabled.args = untoggledDisabledProps;

const toggledProps = {
	...standardProps,
	isSelected: true,
};

export const toggled: ComponentStory<typeof Switch> =
	Template.bind(toggledProps);
toggled.args = toggledProps;

const toggledDisabledProps = {
	...standardProps,
	isSelected: true,
	isDisabled: true,
};

export const toggledDisabled: ComponentStory<typeof Switch> =
	Template.bind(toggledDisabledProps);
toggledDisabled.args = toggledDisabledProps;
