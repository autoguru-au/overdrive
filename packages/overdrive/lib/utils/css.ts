const cssVarRegExp = /var\(([^)]+)\)/;

export const cssVarUnwrap = (value: string) => {
	const matches = cssVarRegExp.exec(value);

	return matches ? matches[1] : value;
};

export const getThemeTokenValue = (
	themeClass: string,
	token: string,
): string | null => {
	const cssVar = cssVarUnwrap(token);
	const themedElement = document.querySelector(`.${themeClass}`);
	if (!themedElement || !cssVar) return null;
	return (
		getComputedStyle(themedElement).getPropertyValue(cssVar)?.trim() || null
	);
};
