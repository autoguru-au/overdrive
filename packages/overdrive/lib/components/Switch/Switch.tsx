import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, memo, useCallback, useState } from 'react';
import { useStyles } from 'react-treat';

import { Box, useBoxStyles } from '../Box';
import { useTextStyles } from '../Text';

import * as styleRefs from './Switch.treat';

export interface Props {
	className?: string;
	disabled?: boolean;
	toggled?: boolean;

	onChange?(value: boolean): void;
}

const SwitchComponent: FunctionComponent<Props> = ({
	className = '',
	disabled = false,
	toggled: incomingToggled = false,
	onChange,
}) => {
	const styles = useStyles(styleRefs);
	const [toggled, setToggled] = useState<boolean>(incomingToggled);

	const [prevValue, setPrevValue] = useState<boolean>(incomingToggled);

	const onToggle = useCallback(() => {
		if (disabled) {
			return;
		}

		if (typeof onChange === 'function') {
			onChange(!toggled);
		}

		setToggled(!toggled);
	}, [incomingToggled, disabled, toggled]);

	if (prevValue !== incomingToggled) {
		setToggled(incomingToggled);
		setPrevValue(incomingToggled);
	}

	return (
		<Box
			is="button"
			className={clsx(
				styles.root,
				useTextStyles({ size: '5' }),
				{
					[styles.toggled]: toggled,
					[styles.disabled.default]: disabled,
					[styles.disabled.toggled]: toggled && disabled,
				},
				className,
			)}
			borderRadius="pill"
			position="relative"
			aria-disabled={disabled}
			aria-label={`toggle ${toggled ? 'on' : 'off'}`}
			onClick={onToggle}>
			<Box
				borderWidth="1"
				position="absolute"
				borderRadius="pill"
				backgroundColour="white"
				padding="none"
				boxShadow="2"
				className={clsx(
					styles.handle.default,
					useBoxStyles({ is: 'button' }),
					useTextStyles({ colour: 'white' }),
					{
						[styles.handle.transition]: toggled,
					},
				)}
			/>
		</Box>
	);
};

export const Switch = memo(SwitchComponent);
