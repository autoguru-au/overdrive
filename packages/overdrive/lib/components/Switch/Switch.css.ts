import { style, styleVariants } from '@vanilla-extract/css';

import { themeContractVars as vars } from '../../themes/theme.css';

const handleSize = '24px';
const handleOffset = '2px';
const borderSize = '1px';

export const handle = styleVariants({
	default: {
		borderColor: vars.border.colours.gray,
		top: `calc(${handleOffset} - ${borderSize})`,
		left: `calc(${handleOffset} - ${borderSize})`,
		width: `${handleSize}`,
		height: `${handleSize}`,
		transition: 'transform 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
		willChange: 'transform',
	},
	transition: {
		transform: `translateX(calc(${handleSize} - (2 * ${handleOffset})))`,
		borderColor: vars.colours.intent.primary.background.strong,
	},
});

export const toggled = style({
	borderColor: vars.colours.intent.primary.background.strong,
	backgroundColor: vars.colours.intent.primary.background.standard,
});

export const disabled = styleVariants({
	default: {
		selectors: {
			'&[aria-disabled=true]': {
				cursor: 'not-allowed',
				borderColor: vars.border.colours.light,
				backgroundColor: vars.border.colours.light,
			},
		},
	},
	toggled: {
		selectors: {
			'&[aria-disabled=true]': {
				borderColor: vars.border.colours.light,
				backgroundColor: vars.colours.intent.primary.background.mild,
			},
		},
	},
});

export const root = style({
	width: `calc((2 * ${handleSize}) + ${handleOffset} - 2 * ${borderSize})`,
	height: `calc(${handleSize} + (${handleOffset} * 2))`,
	transition: 'background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
	border: `${borderSize} solid ${vars.border.colours.light}`,
	selectors: {
		[`&:not(${disabled.default}):not(${disabled.toggled}):focus`]: {
			borderColor: vars.colours.intent.information.background.standard,
		},
	},
});
