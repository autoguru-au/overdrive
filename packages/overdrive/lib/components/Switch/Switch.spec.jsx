import { act, fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { useState } from 'react';
import * as styles from './Switch.css';

import { Switch } from './Switch';

describe('<Switch />', () => {
	it('should not throw', () =>
		expect(() => render(<Switch />)).not.toThrow());

	it('should match snapshot without props', () => {
		expect(render(<Switch />).container.firstChild).toMatchSnapshot();
	});

	it('should match snapshot when un-toggled', () => {
		expect(
			render(<Switch toggled={false} />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when un-toggled and disabled', () => {
		expect(
			render(<Switch disabled toggled={false} />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when toggled', () => {
		expect(
			render(<Switch toggled />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when toggled and disabled', () => {
		expect(
			render(<Switch toggled disabled />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should pass on className to dom element', () => {
		expect(
			render(<Switch className="toggleButton-class" value={10} />)
				.container.firstChild,
		).toHaveClass('toggleButton-class');
	});

	it('should set toggle to false by default', () => {
		expect(render(<Switch />).container.firstChild).not.toHaveClass(
			styles.toggled,
		);
	});

	it('should be toggled on when toggled prop is set to true', () => {
		expect(render(<Switch toggled />).container.firstChild).toHaveClass(
			styles.toggled,
		);
	});

	it('should be enabled by default', () => {
		expect(
			render(<Switch />).container.querySelector('.disabled'),
		).not.toBeInTheDocument();
	});

	it('should have aria-disabled attribute when disabled', () => {
		const { container } = render(<Switch disabled />);

		expect(container.firstChild).toHaveClass(styles.disabled.default);
		expect(container.firstChild).toHaveAttribute('aria-disabled', 'true');
	});

	it('should have aria-disabled attribute when toggled and disabled', () => {
		const { container } = render(<Switch toggled disabled />);

		expect(container.firstChild).toHaveClass(styles.disabled.toggled);
		expect(container.firstChild).toHaveAttribute('aria-disabled', 'true');
	});

	it('should fire change with the correct changed value when clicked', () => {
		const spyedCallback = jest.fn();

		const { container } = render(
			<Switch toggled={false} onChange={spyedCallback} />,
		);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(true);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(false);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(true);
		expect(spyedCallback).toHaveBeenCalledTimes(3);
	});

	it('should not fire change if clicked while disabled', () => {
		const spyedCallback = jest.fn();

		const { container } = render(
			<Switch disabled toggled={false} onChange={spyedCallback} />,
		);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).not.toHaveBeenCalled();
	});

	it('should update its value when and a value prop comes in', () => {
		const ToggleButtonWrapper = ({ setter }) => {
			const [toggled, toggledValue] = useState(false);

			setter(toggledValue);

			return <Switch toggled={toggled} />;
		};

		let setToggledValue;
		const { container } = render(
			<ToggleButtonWrapper
				setter={(setter) => {
					setToggledValue = setter;
				}}
			/>,
		);

		expect(container.firstChild).not.toHaveClass(styles.toggled);

		act(() => setToggledValue(true));

		expect(container.firstChild).toHaveClass(styles.toggled);

		act(() => setToggledValue(false));

		expect(container.firstChild).not.toHaveClass(styles.toggled);
	});
});
