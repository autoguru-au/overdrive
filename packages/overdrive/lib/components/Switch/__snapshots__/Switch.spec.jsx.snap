// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should match snapshot when toggled 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle on"
  class="base root_base toggled_base"
>
  <span
    class="handle_default_base handle_transition_base"
  />
</div>
`;

exports[`<Switch /> should match snapshot when toggled and disabled 1`] = `
<div
  aria-disabled="true"
  aria-label="toggle on"
  class="base root_base toggled_base disabled_default_base disabled_toggled_base"
>
  <span
    class="handle_default_base handle_transition_base"
  />
</div>
`;

exports[`<Switch /> should match snapshot when un-toggled 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle off"
  class="base root_base"
>
  <span
    class="handle_default_base"
  />
</div>
`;

exports[`<Switch /> should match snapshot when un-toggled and disabled 1`] = `
<div
  aria-disabled="true"
  aria-label="toggle off"
  class="base root_base disabled_default_base"
>
  <span
    class="handle_default_base"
  />
</div>
`;

exports[`<Switch /> should match snapshot without props 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle off"
  class="base root_base"
>
  <span
    class="handle_default_base"
  />
</div>
`;
